lisa/sut_orchestrator/libvirt/platform.py

```diff
@@ -709,13 +722,74 @@ def _fill_nodes_metadata(self, environment: Environment, log: Logger) -> None:
                 private_key_file=self.runbook.admin_private_key_file,
             )
 
-            # Ensure cloud-init completes its setup.
-            node.execute(
-                "cloud-init status --wait",
-                sudo=True,
-                expected_exit_code=0,
-                expected_exit_code_failure_message="waiting on cloud-init",
+            node_context = get_node_context(node)
+            if node_context.use_cloud_init:
+                # Ensure cloud-init completes its setup.
+                node.execute(
+                    "cloud-init status --wait",
+                    sudo=True,
+                    expected_exit_code=0,
+                    expected_exit_code_failure_message="waiting on cloud-init",
+                )
+
+    # Setup Ignition for a VM.
+    def _create_node_ignition(
+        self, environment: Environment, log: Logger, node: Node
+    ) -> None:
+        environment_context = get_environment_context(environment)
+        node_context = get_node_context(node)
+
+        user_data = {
+           "ignition": {
+               "version": "3.3.0",
+           },
+           "passwd": {
+               "users": [
+                   {
+                       "name": self.runbook.admin_username,
+                       "sshAuthorizedKeys": [environment_context.ssh_public_key],
+                   },
+               ],
+           },
+        }
+
+        # Iterate through all the top-level properties.
+        for extra_user_data in node_context.extra_cloud_init_user_data:
```

Comment:
The cloud-init data structure is fairly flat. So, this simple merge scheme works well for most things. But it looks like Ignition schema is a little more nested. For example, the `users` list sits under the `passwd` property. So, the merge scheme might not be particularly helpful.

Do you actually have a use case for merging user provided values? If not, it might be worth removing this for now.